File: rust/do/src/nightup/odin.rs
use regex::Regex;
use std::fs;
use std::path::Path;
use std::process::{Command, ExitCode};

pub fn run(dist_dir: &str, download_dir: &str) -> ExitCode {
    // index.jsonを取得する
    let output = match Command::new("curl")
        .args([
            "-sSL",
            "-A",
            "Mozilla/5.0",
            "https://f001.backblazeb2.com/file/odin-binaries/nightly.json",
        ])
        .output()
    {
        Ok(out) => out,
        Err(_) => {
            eprintln!("failed to 'curl'");
            return ExitCode::FAILURE;
        }
    };
    println!("Download (nightly.json) is done");

    let contents = match String::from_utf8(output.stdout) {
        Ok(contens) => contens,
        Err(_) => {
            eprintln!("failed to 'String::from_utf8'");
            return ExitCode::FAILURE;
        }
    };

    // ダウンロードURLを取得する
    let re_date = Regex::new(r#""([\d]{4}-[\d]{2}-[\d]{2})T"#)
        .expect("failed to compile odin.re_date (regex)");

    // 抽出できた URL を格納する変数
    let mut nightly_date = String::new();

    if let Some(caps) = re_date.captures(&contents) 
        && let Some(mat) = caps.get(1) {
            nightly_date = mat.as_str().to_string();
    }

    if nightly_date.is_empty() {
        eprintln!("failed to find ZIP URL for odin-windows-amd64 nightly");
        return ExitCode::FAILURE;
    }

    let zip_name = format!("odin-windows-amd64-nightly%2B{}.zip", nightly_date);
    let download_url = format!(
        "https://f001.backblazeb2.com/file/odin-binaries/nightly/{}",
        zip_name
    );
    println!("Download URL: {}", download_url);

    let work_dir_name = "odin-nightly-upgrade-working";
    let work_dir_path = Path::new(download_dir).join(work_dir_name);

    if work_dir_path.exists() {
        if let Err(e) = fs::remove_dir_all(&work_dir_path) {
            eprintln!("failed to remove {}: {}", work_dir_path.display(), e);
            return ExitCode::FAILURE;
        }
        println!(r#"Removed: "{}""#, work_dir_path.display());
    }

    if let Err(e) = fs::create_dir_all(&work_dir_path) {
        eprintln!("failed to create {}: {}", work_dir_path.display(), e);
        return ExitCode::FAILURE;
    }
    println!(r#"Created: "{}""#, work_dir_path.display());

    let local_zip = "odin-nightly-latest.zip";

    let output = match Command::new("curl")
        .current_dir(&work_dir_path)
        .args(["-fsSL", "-A", "Mozilla/5.0", &download_url, "-o", local_zip])
        .output()
    {
        Ok(out) => out,
        Err(_) => {
            eprintln!("failed to 'curl ZIP file'");
            return ExitCode::FAILURE;
        }
    };

    if !output.status.success() {
        eprintln!("curl failed with status: {:?}", output.status);
        return ExitCode::FAILURE;
    }
    println!("Download (ZIP) is done");

    let output = match Command::new("tar")
        .current_dir(&work_dir_path)
        .args(["-xf", local_zip, "--strip-components=1"])
        .output()
    {
        Ok(out) => out,
        Err(_) => {
            eprintln!("failed to run OS native 'tar'. Make sure Windows 10/11 is updated.");
            return ExitCode::FAILURE;
        }
    };

    if !output.status.success() {
        eprintln!("tar failed with status: {:?}", output.status);
        return ExitCode::FAILURE;
    }
    println!("Extraction is done");

    // 後始末と配置
    let zip_path = work_dir_path.join(local_zip);
    if zip_path.exists() {
        if let Err(e) = fs::remove_file(&zip_path) {
            eprintln!("failed to remove {}: {}", zip_path.display(), e);
            return ExitCode::FAILURE;
        }
        println!(r#"Removed: "{}""#, zip_path.display());
    }

    // 配置（アップデートの適用）
    let target_path = Path::new(dist_dir);
    // 既存の古いインストールディレクトリがあれば削除
    if target_path.exists() {
        if let Err(e) = fs::remove_dir_all(target_path) {
            eprintln!("failed to remove old dist directory: {}", e);
            return ExitCode::FAILURE;
        }
        println!(r#"Removed: "{}""#, target_path.display());
    }

    // 中身がファイル群だけになったワークスペースそのものを、そのまま dist_dir のパスへ移動リネームする
    if let Err(e) = fs::rename(&work_dir_path, target_path) {
        eprintln!("failed to move extracted directory to dist: {}", e);
        return ExitCode::FAILURE;
    }
    println!(
        r#"Moved: "{}" to "{}""#,
        work_dir_path.display(),
        target_path.display()
    );

    println!(r#"Updated: "{}""#, dist_dir);

    ExitCode::SUCCESS
}

File: nim/do/src/nightup/odin.nim
import std/[os, osproc, strformat]
import regex

import ../[utils]

proc run*(distDir: string, downloadDir: string) =
  # 1. 最新バージョンのJSONを取得
  let url = "https://f001.backblazeb2.com/file/odin-binaries/nightly.json"
  let (output, curlRes) = execCmdEx(fmt"""curl -sSL -A "Mozilla/5.0" "{url}"""")
  if curlRes != 0:
    stderrMsgAndExit "failed to download nightly.json"
  
  echo "Download (nightly.json) is done"

  # 2. 正規表現で日付（YYYY-MM-DD）を抽出
  let pattern = re2("""([\d]{4}-[\d]{2}-[\d]{2})T""")
  var m = RegexMatch2()
  let match = find(output, pattern, m)

  var nightlyDate = ""
  if match:
    nightlyDate = output[m.group(0)]

  if nightlyDate == "":
    stderrMsgAndExit "failed to find ZIP URL for odin-windows-amd64 nightly"

  # URLエンコードされた「+」である「%2B」を使用してZIP名とURLを構築
  let zipName = "odin-windows-amd64-nightly%2B" & nightlyDate & ".zip"
  let downloadUrl = "https://f001.backblazeb2.com/file/odin-binaries/nightly/" & zipName
  echo "Download URL: ", downloadUrl

  # 3. 作業用ディレクトリの作成
  let workDirName = "odin-nightly-upgrade-working"
  let workDirPath = downloadDir / workDirName

  try:
    rmDirIfExist(workDirPath)
    echo fmt"""Removed: "{workDirPath}""""
  except CatchableError:
    stderrMsgAndExit fmt"failed to remove existing work dir: {workDirPath}"

  try:
    createDir(workDirPath)
    echo fmt"""Created: "{workDirPath}""""
  except CatchableError:
    stderrMsgAndExit fmt"failed to create work dir: {workDirPath}"

  # 4. ZIPファイルのダウンロード
  let localZip = "odin-nightly-latest.zip"
  let localZipPath = workDirPath / localZip

  let zipProcess = startProcess(
    "curl",
    args = ["-fsSL", "-A", "Mozilla/5.0", downloadUrl, "-o", localZip],
    workingDir = workDirPath,
    options = {poUsePath, poParentStreams}
  )
  let zipExit = zipProcess.waitForExit()
  zipProcess.close()

  if zipExit != 0:
    rmDirIfExist(workDirPath)
    stderrMsgAndExit "failed to download ZIP"
  
  echo "Download (ZIP) is done"

  # 5. 外部コマンド tar の実行
  let tarProcess = startProcess(
    "tar",
    args = ["-xf", localZip, "--strip-components=1"],
    workingDir = workDirPath,
    options = {poUsePath, poParentStreams}
  )
  let tarExit = tarProcess.waitForExit()
  tarProcess.close()

  if tarExit != 0:
    rmDirIfExist(workDirPath)
    stderrMsgAndExit "failed to extract ZIP"
  
  echo "Extraction is done"

  # 6. 不要になったZIPの削除
  try:
    rmIfExist(localZipPath)
    echo fmt"""Removed: "{localZipPath}""""
  except CatchableError:
    rmDirIfExist(workDirPath)
    stderrMsgAndExit "failed to remove local ZIP file"

  # 7. 配置（アップデートの適用）
  try:
    rmDirIfExist(distDir)
    echo fmt"""Removed: "{distDir}""""
  except CatchableError:
    rmDirIfExist(workDirPath)
    stderrMsgAndExit fmt"failed to remove distDir: {distDir}"

  # ワークスペースを作業パスから distDir へ移動
  try:
    moveDir(workDirPath, distDir)
    echo fmt"""Moved: "{workDirPath}" to "{distDir}""""
    echo fmt"""Updated: "{distDir}""""
  except CatchableError:
    rmDirIfExist(workDirPath)
    stderrMsgAndExit fmt"failed to move directory to {distDir}"
